web-api/src/personalization/filter.rs

```diff
@@ -147,10 +148,142 @@ impl<'de> Deserialize<'de> for Compare {
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: SeqAccess<'de>,
+            {
+                let size = seq.size_hint().unwrap_or_default();
+                if size > Self::Value::MAX_LEN {
+                    return Err(A::Error::invalid_length(size, &Self));
+                }
+
+                let mut filters = Vec::with_capacity(size);
+                while let Some(filter) = seq.next_element()? {
+                    if filters.len() < Self::Value::MAX_LEN {
+                        filters.push(filter);
+                    } else {
+                        return Err(A::Error::invalid_length(filters.len() + 1, &Self));
+                    }
+                }
+
+                Ok(Filters(filters))
+            }
+        }
+
+        deserializer.deserialize_seq(FiltersVisitor)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct Combine {
+    pub(crate) operation: CombineOp,
+    pub(crate) filters: Filters,
+}
+
+impl Combine {
+    const MAX_LEVEL: usize = 2;
+}
+
+impl<'de> Deserialize<'de> for Combine {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct CombineVisitor;
+
+        impl<'de> Visitor<'de> for CombineVisitor {
+            type Value = Combine;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str(&format!(
+                    "a json object with exactly one combination operator and at most {} times nested combinations",
+                    Self::Value::MAX_LEVEL,
+                ))
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: MapAccess<'de>,
+            {
+                let Some((operation, filters)) = map.next_entry::<_, Filters>()? else {
+                    return Err(A::Error::invalid_length(0, &Self));
+                };
+                if map.next_entry::<CombineOp, Vec<Filter>>()?.is_some() {
+                    return Err(A::Error::invalid_length(
+                        2 + map.size_hint().unwrap_or_default(),
+                        &Self,
+                    ));
+                }
+                if !filters.is_below_level(Self::Value::MAX_LEVEL) {
+                    return Err(A::Error::invalid_value(
+                        Unexpected::Other(&format!(
+                            "more than {} times nested combinations",
+                            Self::Value::MAX_LEVEL,
+                        )),
+                        &Self,
+                    ));
+                }
+
+                Ok(Combine { operation, filters })
+            }
+        }
+
+        deserializer.deserialize_map(CombineVisitor)
+    }
+}
+
 #[derive(Clone, Debug, Deserialize, PartialEq)]
 #[serde(untagged)]
 pub(crate) enum Filter {
     Compare(Compare),
+    Combine(Combine),
```

Comment:
it's not about performance but about making sure we never parse `$and` or `$or` accidentally as field names.

Like when adding `$document.tags` or similar .

Reply:
from a user input this can never happen as we don't allow `$` as a char in ids. if we on our side decide to add some special ids using forbidden chars, then we need to very carefully evaluate if and what to use, we probably need a list of reserved special ids then. until then we can keep it as is.